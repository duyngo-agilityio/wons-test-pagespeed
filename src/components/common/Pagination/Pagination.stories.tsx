import { Meta, StoryObj } from '@storybook/react';

// components
import Pagination from './index';

const meta: Meta<typeof Pagination> = {
  title: 'Components/common/Pagination',
  component: Pagination,
  argTypes: {
    total: {
      description: 'Total number of pages',
    },
  },
  args: {
    total: 10,
  },
  parameters: {
    controls: {
      expanded: true,
    },
  },
};

export default meta;
type Story = StoryObj<typeof Pagination>;

export const Primary: Story = {};
