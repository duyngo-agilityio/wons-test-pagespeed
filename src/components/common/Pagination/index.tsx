'use client';

import { useCallback } from 'react';
import {
  extendVariants,
  Pagination as NextUIPagination,
  PaginationProps as NextUIPaginationProps,
} from '@nextui-org/react';
import { usePathname, useRouter, useSearchParams } from 'next/navigation';

// themes
import { colors } from '@/themes';

// Constants
import { DEFAULT_PAGE, SEARCH_QUERIES } from '@/constants';

export const CustomPagination = extendVariants(NextUIPagination, {
  variants: {
    color: {
      primary: {
        item: 'text-foreground hover:bg-primary hover:bg-opacity-25',
        wrapper: 'text-white',
        cursor: `bg-[${colors.blue[500]}]`,
      },
    },
  },

  defaultVariants: {
    color: 'primary',
  },
});

const Pagination = ({ ...props }: NextUIPaginationProps) => {
  const pathname = usePathname();
  const searchParams = useSearchParams();
  const { replace } = useRouter();

  const currentPage = searchParams?.get(SEARCH_QUERIES.PAGE) || DEFAULT_PAGE;

  const handlePageChange = useCallback(
    (page: number) => {
      const params = new URLSearchParams(searchParams);

      params.set(SEARCH_QUERIES.PAGE, page.toString());

      replace(`${pathname}?${params.toString()}`, { scroll: false });
    },
    [pathname, replace, searchParams],
  );

  return (
    <CustomPagination
      page={+currentPage}
      onChange={handlePageChange}
      className="flex justify-center m-0"
      {...props}
    />
  );
};

export default Pagination;
