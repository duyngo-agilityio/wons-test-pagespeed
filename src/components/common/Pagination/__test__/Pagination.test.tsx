// Libs
import { PaginationProps } from '@nextui-org/react';

// Components
import Pagination from '../index';

jest.mock('next/navigation', () => ({
  __esModule: true,
  useRouter() {
    return {
      prefetch: () => null,
    };
  },
  useSearchParams: () => ({
    get: (key: string) => {
      if (key === 'page') {
        return undefined;
      }
      return null;
    },
  }),
  usePathname: () => '/mock-path',
}));

describe('Pagination test cases', () => {
  const renderComponent = (props: PaginationProps) =>
    testLibJestUtils.render(<Pagination {...props} />);

  it('should render correctly', () => {
    const { container } = renderComponent({
      total: 5,
    });

    expect(container).toMatchSnapshot();
  });
});
