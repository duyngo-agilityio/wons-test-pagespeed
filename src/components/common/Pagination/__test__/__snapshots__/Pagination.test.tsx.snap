// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination test cases should render correctly 1`] = `
<div>
  <div
    data-overlay-container="true"
  >
    <script>
      !function(){try{var d=document.documentElement,c=d.classList;c.remove('light','dark');var e=localStorage.getItem('theme');if('system'===e||(!e&&false)){var t='(prefers-color-scheme: dark)',m=window.matchMedia(t);if(m.media!==t||m.matches){d.style.colorScheme = 'dark';c.add('dark')}else{d.style.colorScheme = 'light';c.add('light')}}else if(e){c.add(e|| '')}else{c.add('light')}if(e==='light'||e==='dark'||!e)d.style.colorScheme=e||'light'}catch(e){}}()
    </script>
     
    <nav
      aria-label="pagination navigation"
      class="p-2.5 overflow-x-scroll scrollbar-hide flex justify-center m-0"
      data-active-page="1"
      data-dots-jump="5"
      data-slot="base"
      data-total="5"
      role="navigation"
    >
      <ul
        class="flex flex-nowrap h-fit max-w-fit relative gap-1 items-center overflow-visible rounded-medium text-white"
        data-slot="wrapper"
      >
        <span
          aria-hidden="true"
          class="absolute flex overflow-visible items-center justify-center origin-center left-0 select-none touch-none pointer-events-none z-20 text-primary-foreground data-[moving=true]:transition-transform !data-[moving=true]:duration-300 opacity-0 data-[moving]:opacity-100 min-w-9 w-9 h-9 text-small rounded-medium bg-[#3A36DB]"
          data-moving="false"
          data-slot="cursor"
          style="transform: translateX(0px) scale(1);"
        >
          1
        </span>
        <li
          aria-current="true"
          aria-label="pagination item 1 active"
          class="tap-highlight-transparent select-none touch-none data-[pressed=true]:scale-[0.97] transition-transform-background flex flex-wrap truncate box-border items-center justify-center outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 data-[disabled=true]:text-default-300 data-[disabled=true]:pointer-events-none shadow-sm bg-default-100 [&[data-hover=true]:not([data-active=true])]:bg-default-200 active:bg-default-300 min-w-9 w-9 h-9 text-small rounded-medium text-foreground hover:bg-primary hover:bg-opacity-25"
          data-active="true"
          data-slot="item"
          role="button"
          tabindex="0"
        >
          1
        </li>
        <li
          aria-label="pagination item 2"
          class="tap-highlight-transparent select-none touch-none data-[pressed=true]:scale-[0.97] transition-transform-background flex flex-wrap truncate box-border items-center justify-center outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 data-[disabled=true]:text-default-300 data-[disabled=true]:pointer-events-none shadow-sm bg-default-100 [&[data-hover=true]:not([data-active=true])]:bg-default-200 active:bg-default-300 min-w-9 w-9 h-9 text-small rounded-medium text-foreground hover:bg-primary hover:bg-opacity-25"
          data-slot="item"
          role="button"
          tabindex="0"
        >
          2
        </li>
        <li
          aria-label="pagination item 3"
          class="tap-highlight-transparent select-none touch-none data-[pressed=true]:scale-[0.97] transition-transform-background flex flex-wrap truncate box-border items-center justify-center outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 data-[disabled=true]:text-default-300 data-[disabled=true]:pointer-events-none shadow-sm bg-default-100 [&[data-hover=true]:not([data-active=true])]:bg-default-200 active:bg-default-300 min-w-9 w-9 h-9 text-small rounded-medium text-foreground hover:bg-primary hover:bg-opacity-25"
          data-slot="item"
          role="button"
          tabindex="0"
        >
          3
        </li>
        <li
          aria-label="pagination item 4"
          class="tap-highlight-transparent select-none touch-none data-[pressed=true]:scale-[0.97] transition-transform-background flex flex-wrap truncate box-border items-center justify-center outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 data-[disabled=true]:text-default-300 data-[disabled=true]:pointer-events-none shadow-sm bg-default-100 [&[data-hover=true]:not([data-active=true])]:bg-default-200 active:bg-default-300 min-w-9 w-9 h-9 text-small rounded-medium text-foreground hover:bg-primary hover:bg-opacity-25"
          data-slot="item"
          role="button"
          tabindex="0"
        >
          4
        </li>
        <li
          aria-label="pagination item 5"
          class="tap-highlight-transparent select-none touch-none data-[pressed=true]:scale-[0.97] transition-transform-background flex flex-wrap truncate box-border items-center justify-center outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 data-[disabled=true]:text-default-300 data-[disabled=true]:pointer-events-none shadow-sm bg-default-100 [&[data-hover=true]:not([data-active=true])]:bg-default-200 active:bg-default-300 min-w-9 w-9 h-9 text-small rounded-medium text-foreground hover:bg-primary hover:bg-opacity-25"
          data-slot="item"
          role="button"
          tabindex="0"
        >
          5
        </li>
      </ul>
    </nav>
    <div
      class="fixed top-4 right-4"
    />
  </div>
</div>
`;
